protocols/upnp/src/behaviour.rs

```diff
@@ -0,0 +1,554 @@
+// Copyright 2023 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+#![cfg_attr(docsrs, feature(doc_cfg, doc_auto_cfg))]
+
+use std::{
+    borrow::Borrow,
+    collections::{HashMap, VecDeque},
+    error::Error,
+    hash::{Hash, Hasher},
+    marker::PhantomData,
+    net::SocketAddrV4,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use crate::{
+    provider::{Gateway, Protocol, Provider},
+    Config,
+};
+use futures::{future::BoxFuture, Future, FutureExt, StreamExt};
+use futures_timer::Delay;
+use libp2p_core::{multiaddr, transport::ListenerId, Endpoint, Multiaddr};
+use libp2p_swarm::{
+    derive_prelude::PeerId, dummy, ConnectionDenied, ConnectionId, ExpiredListenAddr, FromSwarm,
+    NetworkBehaviour, NewListenAddr, PollParameters, ToSwarm,
+};
+
+/// The duration in seconds of a port mapping on the gateway.
+const MAPPING_DURATION: u32 = 3600;
+
+/// Renew the Mapping every half of `MAPPING_DURATION` to avoid the port being unmapped.
+const MAPPING_TIMEOUT: u64 = MAPPING_DURATION as u64 / 2;
+
+/// A [`Gateway`] Request.
+#[derive(Debug)]
+pub(crate) enum GatewayRequest {
+    AddMapping(Mapping, Option<u32>),
```

Comment:
```suggestion
    AddMapping {
        mapping: Mapping,
        whatever_this_is: Option<u32>,
    },
```

I suggest giving the `u32` a name.

Reply:
makes sense, updated!